alerting/src/main/kotlin/org/opensearch/alerting/util/AggregationQueryRewriter.kt

```diff
@@ -82,11 +87,37 @@ class AggregationQueryRewriter {
                     val lastAgg = aggs.asMap[parentBucketPath.pathElements.last().name]
                     // if leaf is CompositeAggregation, then fetch afterKey if present
                     if (lastAgg is CompositeAggregation) {
-                        aggTriggerAfterKeys[trigger.id] = lastAgg.afterKey()
+                        /*
+                         * Bucket-Level Triggers can have different parent bucket paths that they are tracking for condition evaluation.
+                         * These different bucket paths could have different page sizes, meaning one could be exhausted while another
+                         * bucket path still has pages to iterate in the query responses.
+                         *
+                         * To ensure that these can be exhausted and tracked independently, the after key that led to the last page (which
+                         * should be an empty result for the bucket path) will be saved when the last page is hit and will be continued
+                         * to be passed on for that bucket path if there are still other bucket paths being paginated.
+                         */
+                        val afterKey = lastAgg.afterKey()
+                        val prevTriggerAfterKey = prevBucketLevelTriggerAfterKeys?.get(trigger.id)
+                        if (prevTriggerAfterKey == null) {
+                            // If the previous TriggerAfterKey was null, this should be the first page
+                            bucketLevelTriggerAfterKeys[trigger.id] = TriggerAfterKey(afterKey, afterKey == null)
+                        } else {
+                            if (prevTriggerAfterKey.lastPage) {
+                                // If the previous TriggerAfterKey already hit the last page, pass along the after key it used to get there
+                                bucketLevelTriggerAfterKeys[trigger.id] = prevTriggerAfterKey
+                            } else if (afterKey == null) {
+                                // If the previous TriggerAfterKey had not reached the last page and the after key for the current result
+                                // is null, then the last page has been reached so the after key that was used to get there is stored
+                                bucketLevelTriggerAfterKeys[trigger.id] = TriggerAfterKey(prevTriggerAfterKey.afterKey, true)
+                            } else {
+                                // Otherwise, update the after key to the current one
+                                bucketLevelTriggerAfterKeys[trigger.id] = TriggerAfterKey(afterKey, false)
```

Comment:
Do you think we should simplify and cleanup these if else blocks. Couple of solutions which come to my mind are abstracting it out in a utils function or maybe using a switch case or maybe ternary operators. I leave it upto you to make the decision.

Reply:
Updated it to a `when` statement